Error recovery, node reuse, multinodes

ptersilie opened the pull request:
This PR will merge multiple branches I've been working on into master:

Error recovery: After finding an error the parser can continue to parse the rest of the program and merge successfully parsed subtrees into the current parse tree. Includes out-of-context-analysis which can parse subtrees after the error node within the isolated subtree, and (currently disabled) retainability which integrates subtrees within the iso tree before the error node.

Node reuse: When reducing nodes into a new nonterminal this feature enables the reuse of the former parent of one of the nodes. This avoids recreating nonterminal nodes that are already in the parse tree.

Multinodes: Enables the lexer to create Terminal nodes that consists of multiple other terminals. This is needed to implement language boxes and newlines within strings and comments.

ptersilie committed:
Refactored versioning and change reporting

ptersilie committed:
Refactored history services and undo/redo

All tests passing

ptersilie committed:
This test fails for some reason (deepcopy related?)

ptersilie committed:
Added node reuse and refactored python indentations to work with it

ptersilie committed:
Implemented basic error recovery

Doesn't reevaluate isolation trees whose surrounding context has changed.
Doesn't find nested errors within already isolated trees.

ptersilie committed:
Fixed some bugs and added color for isolated nodes

- Isolated attribute was reset straight after setting it
- stack_offset returned the wrong offset, leading to bigger isolation trees

ptersilie committed:
Added tests and fixed bugs

ptersilie committed:
Cleanup and comments

ptersilie committed:
Fix cursor movement and relations between MultiTextNode children

ptersilie committed:
Avoid relexing nodes that don't need to be relexed.

Also speed optimisations

ptersilie committed:
only break AFTER startnode not ON startnode

ptersilie committed:
Add multinode tests

ptersilie committed:
Update children when removing elements from MultiTextNode

ptersilie committed:
Finish token creation on IndentationTerminals

ptersilie committed:
Track lexing changes to only reparse if needed

ptersilie committed:
cleanup

ptersilie committed:
fix: indentation in first line should produce an error

ptersilie committed:
Added out-of-context recovery

ptersilie committed:
All tests running (OOC deactivated)

ptersilie committed:
Reactivate error recovery, bug fixes and optimisations

Bugs:
- Fixed temporary parser not finishing upon seeing temp_eos
- Ignore deleted nodes when repairing indentations

Optimisations
- Improved get_attr by storing the maximum version of a node
- Calculating textlength everytime the parsetree is saved
- Ignore already checked ancestors in find_iso_tree (like Wagner)

ptersilie committed:
Added retainablity checks and tests

Retaining subtrees is currently disabled due to a bug in the algorithm.

Other changes:
- mark broken down subtrees as deleted
- added exists(version) method

ptersilie committed:
Allow to use debugging in all modes

ptersilie committed:
Mark failing tests with xfail so travis won't fail

ptersilie committed:
First attempt at multiline tokens

# Conflicts:
#	lib/eco/inclexer/inclexer.py
#	lib/eco/incparser/astree.py

ptersilie committed:
Replace MultiTerminal with MultiTextNode

# Conflicts:
#	lib/eco/incparser/astree.py

ptersilie committed:
remove indentation in parser and put back indentationmanager

# Conflicts:
#	lib/eco/incparser/incparser.py

ptersilie committed:
bug fixes and syntax highlighting

# Conflicts:
#	lib/eco/treemanager.py

ptersilie committed:
all tests running with multinodes

ptersilie committed:
Implemented Wagners lookback-update algorithm

# Conflicts:
#	lib/eco/incparser/astree.py

ptersilie committed:
When a LexingError occurs attempt to merge partly matches tokens back into the parse tree.

This commit also continues relexing further changes after a LexingError. This results in some of our fuzzy tests becoming really slow.
So the next step is to restrict the relexing of nodes to those who have changes within their lookahead.

# Conflicts:
#	lib/eco/incparser/astree.py

ptersilie committed:
Merge branch 'master' into history_nodereuse_errorrec_and_multinodes_rebased_2

# Conflicts:
#	lib/eco/inclexer/inclexer.py
#	lib/eco/incparser/incparser.py
#	lib/eco/nodeeditor.py

ltratt commented on lib/eco/cflexer/lexer.py:
```diff
@@ -168,15 +170,18 @@ def find_next_token(self):
                     source = self.text[start:]
-                    result = self.make_token(start, self.last_matched_state, source, lookahead = i - stop)
+                    raise LexingError("Could not successfully parse %s" % source)
+                    result = self.make_token(start, -1, self.last_matched_state, lookahead = i - stop)
```
The rest of this `if`s block almost certainly can't be executed (because of the `raise`).

ptersilie replied:
You are right.

ltratt commented on lib/eco/cflexer/lexer.py:
```diff
@@ -189,13 +194,14 @@ def find_next_token(self):
+                result = self.make_token(start, -1, self.last_matched_state, lookahead = lookahead)
                 self.last_matched_index = start + len(source)
                 self.adjust_position(source)
                 if self.ignore_token(self.last_matched_state):
```
Why not just say `assert not self.eof`.

ptersilie replied:
Ah this was just me testing if these parts are ever reached, which I guess they are not, so I should remove them all together.

ltratt commented on lib/eco/cflexer/lexer.py:
```diff
@@ -244,10 +250,27 @@ def __init__(self, matcher, automaton, text, ignore, eof=False):
-            return Token("EOF", "EOF", source_pos, lookahead)
-        return Token(self.automaton.names[self.last_matched_state],
-                     text, source_pos, lookahead)
+        #source_pos = SourcePos(index, self.lineno, self.columnno)
```
Remove commented line.

ltratt commented on lib/eco/grammar_parser/gparser.py:
```diff
@@ -70,6 +70,8 @@ def __hash__(self):
     def copy(self):
         return self.__class__(self.name, self.folding)
 
+    
```
More blank lines than necessary?

ltratt commented on lib/eco/inclexer/inclexer.py:
```diff
@@ -487,103 +496,299 @@ def relex(self, node):
+
+        combos = []
+        last_read = None
+
```
Not sure this (and the next) blank line are useful?

ltratt commented on lib/eco/incparser/astree.py:
```diff
@@ -231,12 +214,25 @@ def set_children(self, children):
+            if c is self:
+                if i > 0:
+                    c.left = self.parent.children[i-1]
+                if i < len(self.parent.children) - 1:
```
`if` or `elif`?

ptersilie replied:
`if`, since `i` can be between `0` and `len(children)` at the same time.

ltratt commented on lib/eco/incparser/error_recovery.py:
```diff
@@ -0,0 +1,207 @@
+        # Reaching this point means we found a candidate for isolation. Now we
+        # need to check if we can continue parsing if we isolate this subtree
+
+        cut_point = cut
```
There seems to be no need for a new variable `cut_point`? `cut` isn't referenced after this point in the function as far as I can tell.

ptersilie replied:
You are right. Seems to be a left-over from when I used to do something with `cut`.

ltratt commented on lib/eco/incparser/incparser.py:
```diff
@@ -143,62 +141,109 @@ def init_ast(self, magic_parent=None):
         while(True):
             logging.debug("\x1b[35mProcessing\x1b[0m %s %s %s %s", la, la.changed, id(la), la.indent)
             self.loopcount += 1
+
```
Lots of blank lines!

ltratt commented on lib/eco/incparser/incparser.py:
```diff
@@ -143,62 +141,109 @@ def init_ast(self, magic_parent=None):
 
             else: # Nonterminal
-                if la.changed or reparse:
+                #iso_and_changed = la.isolated and self.surrounding_context_changed(la)
```
Commented code.

ltratt commented on lib/eco/incparser/incparser.py:
```diff
@@ -251,7 +310,7 @@ def parse_terminal(self, la, lookup_symbol):
             logging.debug("\x1b[33mReduce\x1b[0m: %s -> %s", la, element)
             self.reduce(element)
-            return self.parse_terminal(la, lookup_symbol)
+            return la #self.parse_terminal(la, lookup_symbol)
```
Commented code.

ltratt commented on lib/eco/incparser/syntaxtable.py:
```diff
@@ -73,15 +73,14 @@ def __init__(self, lr_type=LR0):
             for state in state_set.elements:
                 if state.isfinal():
-                    if state.p == start_production:
+                    if state.p.left == None:
```
`==` -> `is`

ltratt commented on lib/eco/ip_plugins/pythonindentation.py:
```diff
@@ -13,12 +13,68 @@ def incparse_init(self):
+                return self.right_sibling(n)
+            return n
+
+    def wasdeleted(self, node):
```
Is this supposed to be `was_deleted`?

ptersilie replied:
I should be. But the ip_plugins are dead anyways, so there's no need to fix this.

ltratt commented on lib/eco/ip_plugins/pythonindentation.py:
```diff
@@ -13,12 +13,68 @@ def incparse_init(self):
+            c.prev_term = c.prev_term.prev_term
+
+    def pseudo_insert(self, oldnode, newnode):
+        #print("pseudo insert", oldnode.symbol.name, newnode.symbol.name)
```
Commented code.

ltratt commented on lib/eco/ip_plugins/pythonindentation.py:
```diff
@@ -13,12 +13,68 @@ def incparse_init(self):
+        newnode.right = oldnode.right
+        try:
+            print(newnode, "parent", newnode.parent.symbol.name, "changed to ", oldnode.parent.symbol.name)
+        except  AttributeError:
```
Double space after `except`.

ptersilie replied:
How could I have overlooked this? I must have not been myself. Stupid mistake, Lukas, stupid!

ltratt replied:
I was shocked too.

ltratt commented on lib/eco/pgviewer.py:
```diff
@@ -100,16 +100,50 @@ def add_node(self, node, graph):
-            dotnode.set('color','green')
+            pass
+
+        try:
```
These `try` blocks could be squished down into one (since they all share an identical `except` clause).

ptersilie replied:
I don't think they can. If one fails I don't want all the others to fail as well, i.e just because I can't colour mark a node as `changed` doesn't mean it can't be colour-marked as `deleted` or `isolated`.

ltratt replied:
Aha, I see. You're right.

ltratt commented on lib/eco/treemanager.py:
```diff
@@ -1737,44 +1798,74 @@ def export_aterms(self, path):
+        try:
+            return self.relex_node(node)
+        except LexingError:
+            #XXX show LexingError message somwhere in the UI
```
Does this mean that errors are sort-of swallowed?

ptersilie replied:
For now they are, but I was planning to mark them somehow. My initial thought was to also not parse if there is a lexing error, but when we do the partial lexing stuff, then we need to reparse anyway.

ltratt commented on lib/eco/treemanager.py:
```diff
@@ -1737,44 +1798,74 @@ def export_aterms(self, path):
             # later versions are void -> delete
-            self.clean_versions(self.version)
-            self.last_saved_version = self.version
+            #self.clean_versions(self.version)
```
Commented code.

ltratt commented:
OK, so I've made a number of mostly minor suggestions. There are a couple of `if` blocks whose logic I'm not very sure about that might want a slightly deeper check.

ptersilie committed:
Addressed Lauries comments

ptersilie committed:
Enable parsetree history when debugging

ptersilie committed:
Fixed out-of-context-analysis bug

ptersilie commented:
Addressed your comments and also fixed an error recovery bug that one of the fuzzy tests found.

ltratt commented:
These look fine to me. Is this the point we should do some squashing (maybe) so that we can merge? Or ...?

ptersilie commented:
I already did a lot of squashing before raising this PR, but if you have ideas for more squashing let me know.

ltratt commented:
So shall we merge?

ptersilie commented:
Whoops just noticed some typos in my last commits. Let me fix them first.

Also one problem remaining is that the tests on travis time out now (as the slow ones take too long now). How do we fix that?

ltratt commented:
Um, don't know. You may have to select a subset when running under Travis? May be worth reading around...

ptersilie committed:
Travis optimisations

- Cache statetables from previous test runs
- Give the slow runs a little bit more time

ptersilie commented:
Okay, I think this is ready for merging now. I've already got a few more PRs coming up, so there will be more once this is merged.
